public_html/extensions/paypal_commerce/storefront/controller/responses/extension/paypal_commerce.php

```diff
@@ -39,48 +41,48 @@ public function main()
         $this->load->model('checkout/order');
         /** @var ModelExtensionPaypalCommerce $mdl */
         $mdl = $this->load->model('extension/paypal_commerce');
-        $data['client_token'] = $mdl->getClientToken();
-        $data['bn_code'] = $mdl->getBnCode();
-        if (!$data['client_token']) {
-            $data['error'] = 'Cannot to obtain client token from Paypal API. Incident has been reported.';
-            $this->messages->saveError('Paypal Commerce API Error', $data['error'] . "\nSee error log for details.");
+        $this->data['client_token'] = $mdl->getClientToken();
```

Comment:
I want to be able to update the the total amount and products detail via hook for our extension. Are there other way to override than changing the $data to $this->data?

Reply:
no, it's not acceptable.
You cannot to change amounts and product list before order confirmation. 
Your code will change amount, but order summary block will show previous amount.
Amount from order summary and amount on paypal side MUST be equal.

i think you should to create discount total extension's model to change total amount.

Reply:
how admin can make changes to the order before confirmation? 
Is it postponed payment or something like that?

Reply:
yes it is a delayed payment timing.

we're replicating the default bank transfer payment extension so customer still able to place order. other default payment method will be hidden upon placing order. 

After that admin update the order via admin order details,  send the invoice link to customer (storefront page) and customer can select the payment method there then pay.

Reply:
still don't understand how do you see this process. Why you need to change amount on payment controller side? 
You should to change order totals, not payment data

Reply:
I'm thinking what if admin update product qty? Does it get passed to PayPal? 
Also how about if admin update the shipping rates? 
What is your suggestion?

Reply:
In this case admin must change the totals table. This feature allowed for him in admin (see order recalc feature). 
After correction payment extension will have final data which goes from order data (products & totals). 
As i wrote earlier you cannot touch the payment process. This process works only with order data.

Reply:
Hi @abolabo, please review the comment that was posted by @goatlady102.

Thanks.

Reply:
@goatlady102  @handoyo 

To pause the payment process you should to use "override" hook. Please see core/engine/extension.php line 1154 for details.
Example: 
```
public function overrideControllerResponsesExtensionPaypalCommerce_initData()
{
 // confirm order here and set order "pause" status until totals will be tuned.
}
```

Main controller execution will be stopped at initData call point.

Your changes of paypal controller needs to be rollbacked. 
Thank you

Reply:
@abolabo 

I will try to explain our use case again here.

We have create a payment method (clone bank transfer) that allows customer to place an order. 

Here are the flow:
1. customer place order using the cloned bank transfer payment method.
2. admin update the final shipping rates via admin order details page.
3. admin notify customer that they can pay the updated invoice via storefront, customer can select the payment. if i use override, how can i update the new total in paypal commerce so it use the new total?

Reply:
@handoyo  
fyi, you should to override payment controller to replace it's payment form with your custom form to pause order.
Also you can affect on payment list (another approach, probably better than first)

regarding item 3.
Admin notify customer (i guess i will be some link on your custom controller that propose customer to login and push the order_id into his session).
Note: order must be with order status "Incomplete" (order_status_id=0)
After order_id saving into session customer redirected to checkout page and then purchase will be finished as usual.

Reply:
@abolabo 

"fyi, you should to override payment controller to replace it's payment form with your custom form to pause order.
Also you can affect on payment list (another approach, probably better than first)"

I don't understand what do you mean on this. Please explain it.

i have another question about 

`$this->extensions->hk_UpdateData($this, __FUNCTION__);`

in line 322.

How do i hook or override the order total via the hook? I usually hook by using $that->data, but there are no $this->data in the code, but you have the hk_UpdateData?

Reply:
@handoyo 
i suggesting you another way to realize your task. 
1. create hook on fast_checkout controller and check order for changes of total (manual tuning shipping cost). If order have no changes yet - cut off payment list. Leave only your custom payment method which will be payment intent. Let it be just a button "want to buy". After click order will have status "paused". 
fast_checkout session data needs to be saved in the database. 
I think you need to create your own table for that with column type MediumText. The easy way to save  this data is serializing. (serialize($this->session->data['fc']); ). Not sure.. need to try

2. After that merchant can tune the order totals via order total recalc form on order details page. You should add hook on this action (or some another way to inform customer about order is ready to be payoff). 
For example you can add your button "inform customer" for that there.

3. Customer get notification via email with link to your controller. This is special-formed link with hidden parameter order_id (you can take parameter "ot" from account/invoice.php line 48 as example)

4. After click redirect goes to your controller to login (if not-logged). For guests it will be just a redirect. 
After "ot" parameter parsed you paste order_id into his session via
$this->session->data['order_id'] =  $order_id;
and set order status to "incomplete" via sql update query.
Also you should to extract fast checkout data from database (unserialize) and set it as $this->session->data['fc']
and redirect customer to rt checkout/fast_checkout

5. Customer will see the order with edited summary totals and can choose payment method form the list. Than order will be finished as usual by payment method controller.